Domain/EventSourcedRepositoryMigrator{T}.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Copyright (c) Microsoft. All rights reserved. 
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Its.Domain
+{
+    /// <summary>
+    /// If a repository implements the optional <see cref="IMigratableEventSourcedRepository{TAggregate}"/> 
+    /// interface, then this class offers a "side channel" by which clients can rename events in the stored 
+    /// event stream.
+    /// </summary>
+    /// <remarks>Note that existing snapshots and in-memory aggregates are not affected by these migrations.</remarks>
+    public class EventSourcedRepositoryMigrator<TAggregate>
+        where TAggregate : EventSourcedAggregate<TAggregate>
+    {
+        private readonly IMigratableEventSourcedRepository<TAggregate> repository;
+
+        public EventSourcedRepositoryMigrator(IEventSourcedRepository<TAggregate> repository)
+        {
+            if (repository == null)
+            {
+                throw new ArgumentNullException("repository");
+            }
+            this.repository = repository as IMigratableEventSourcedRepository<TAggregate>;
+            if (this.repository == null)
+            {
+                throw new EventSourcedRepositoryMigrator.RepositoryMustSupportMigrationsException(repository.GetType());
+            }
+        }
+
+        public class PendingRenameList : List<Tuple<TAggregate, EventSourcedRepositoryMigrator.RenameRequest>>
+        {
```

Comment:
This is kind of an awkward data structure to be in the public interface, I think. Tuples are a convenience in local code but they tend not to communicate intention very well. If the rename is intended to be associated with a single aggregate instance and no other, why not a class with the rename details as well as a reference to the aggregate on it? Or an aggregate id... Sourcing shouldn't be a requirement for a migration.